okhttp/src/jvmTest/kotlin/okhttp3/TrailersTest.kt

```diff
@@ -0,0 +1,361 @@
+  }
+
+  @Test
+  fun trailersWithoutReadingFullResponseBodyHttp2() {
+    trailersWithoutReadingFullResponseBody(Protocol.HTTP_2)
+  }
+
+  private fun trailersWithoutReadingFullResponseBody(protocol: Protocol) {
+    enableProtocol(protocol)
+
+    server.enqueue(
+      MockResponse
+        .Builder()
+        .addHeader("h1", "v1")
+        .trailers(headersOf("t1", "v2"))
+        .body(protocol, "Hello")
+        .build(),
+    )
+
+    val call = client.newCall(Request(server.url("/")))
+    call.execute().use { response ->
+      assertThat(response.header("h1")).isEqualTo("v1")
+      assertThat(response.trailers()).isEqualTo(headersOf("t1", "v2"))
+    }
+  }
+
+  @Test
+  @Disabled
+  fun trailersAndCacheHttp1() {
+    trailersAndCache(Protocol.HTTP_1_1)
+  }
+
+  @Test
+  @Disabled
+  fun trailersAndCacheHttp2() {
+    trailersAndCache(Protocol.HTTP_2)
+  }
+
+  private fun trailersAndCache(protocol: Protocol) {
+    enableProtocol(protocol)
+
+    server.enqueue(
+      MockResponse
+        .Builder()
+        .addHeader("h1", "v1")
+        .addHeader("Cache-Control: max-age=30")
+        .body(protocol, "Hello")
+        .trailers(headersOf("t1", "v2"))
+        .build(),
+    )
+
+    val call1 = client.newCall(Request(server.url("/")))
+    call1.execute().use { response ->
+      val source = response.body.source()
+      assertThat(response.header("h1")).isEqualTo("v1")
+      assertThat(source.readUtf8()).isEqualTo("Hello")
+      assertThat(response.trailers()).isEqualTo(headersOf("t1", "v2"))
+    }
+
+    val call2 = client.newCall(Request(server.url("/")))
+    call2.execute().use { response ->
+      val source = response.body.source()
+      assertThat(response.header("h1")).isEqualTo("v1")
+      assertThat(source.readUtf8()).isEqualTo("Hello")
+      assertThat(response.trailers()).isEqualTo(headersOf("t1", "v2"))
+    }
+  }
+
+  @Test
+  fun delayBeforeTrailersHttp1() {
+    delayBeforeTrailers(Protocol.HTTP_1_1)
+  }
+
+  @Test
+  fun delayBeforeTrailersHttp2() {
+    trailers(Protocol.HTTP_2)
+  }
+
+  /** Confirm the client will block if necessary to consume trailers. */
+  private fun delayBeforeTrailers(protocol: Protocol) {
+    enableProtocol(protocol)
+
+    server.enqueue(
+      MockResponse
+        .Builder()
+        .addHeader("h1", "v1")
+        .trailers(headersOf("t1", "v2"))
+        .body(protocol, "Hello")
+        .trailersDelay(500, TimeUnit.MILLISECONDS)
+        .build(),
+    )
+
+    val call = client.newCall(Request(server.url("/")))
+    call.execute().use { response ->
+      val source = response.body.source()
+      assertThat(response.header("h1")).isEqualTo("v1")
+      assertThat(source.readUtf8(5)).isEqualTo("Hello")
+      val trailersDelay =
+        measureTime {
+          val trailers = response.trailers()
+          assertThat(trailers).isEqualTo(headersOf("t1", "v2"))
+        }
+      assertThat(trailersDelay).isGreaterThan(250.milliseconds)
```

Comment:
Defending against weirdness like the client assuming there’s no trailers 'cause there’s no trailers yet